pageserver/src/tenant/remote_timeline_client.rs

```diff
@@ -615,6 +628,40 @@ impl RemoteTimelineClient {
         Ok(())
     }
 
+    // NOTE: if there were no tasks to call stop we need to call stop by ourselves first
+    pub(crate) async fn persist_index_part_with_deleted_flag(
+        self: &Arc<Self>,
+    ) -> anyhow::Result<()> {
+        let index_part = {
+            let mut locked = self.upload_queue.lock().unwrap();
+
+            // We must be in stopped state because otherwise
+            // we can have inprogress index part upload that can overwrite the file
+            // with missing is_deleted flag that we going to set below
+            let stopped = match &mut *locked {
+                UploadQueue::Uninitialized => anyhow::bail!("is not Stopped but Uninitialized"),
+                UploadQueue::Initialized(_) => anyhow::bail!("is not Stopped but Initialized"),
+                UploadQueue::Stopped(stopped) => stopped,
+            };
+
+            assert!(
+                stopped.last_uploaded_index_part.deleted_at.is_none(),
+                "deleted_at already set"
+            );
```

Comment:
timeline_delete can run concurrently for the same timeline.
So, we can hit this assert.

Reply:
Fixed this in the commits I posted below, would appreciate review.